theano/tensor/nnet/nnet.py

```diff
@@ -414,40 +477,40 @@ class Softmax(gof.Op):
     :math:`\\varphi(\\mathbf{x})_j =
     \\frac{e^{\mathbf{x}_j}}{\sum_{k=1}^K e^{\mathbf{x}_k}}`
     where :math:`K` is the total number of neurons in the layer. This
-    activation function gets applied row-wise.
-
+    activation function gets applied on the selected axis (Default axis is the last).
+    This function get applied when axis is a scalar and uses a fast C optimization.
     """
 
-    nin = 1
+    nin = 2  # tensor, axis
     nout = 1
     __props__ = ()
+    params_type = ParamsType(axis=Scalar('int8'))
+
+    def __init__(self, axis=-1):
```

Comment:
Yeah, but for this PR, we start by considering only the last axis. It will be easier to write the test for this specific case (Like testing vector, matrix, tensor*d). Then, we can have a new PR that will add this functionality for the user. (Since it will requires to modify the optimization, and that it will generate more cases to test, I prefer to have it in another PR).